### PR TITLE
Use the sharing map for the renaming level maps

### PR DESCRIPTION
**Pull request by danpoe**
Use the sharing map for the renaming level maps `current_names` in `symex_renaming_levelt`. The code in this PR has a few issues which I'll fix once we've decided whether to merge it.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by danpoe** [comment]
Here are some results for the runtime of symex only, with `--unwind 1`, on several methods from Apache Tika on which symex takes between 5 and 15 sec

Rplot
Overall with sharing it's 3% faster.

**Comment by danpoe** [comment]
@smowton, @romainbrenguier Would you be able to try this out on your benchmarks too?

File: src/goto-symex/auto_objects.cpp [hunk 1]
```diff
@@ -91,9 +91,7 @@ void goto_symext::trigger_auto_object(const exprt &expr, statet &state)
       if(has_prefix(id2string(symbol.base_name), "auto_object"))
       {
         // done already?
-        if(
-          state.get_level2().current_names.find(ssa_expr.get_identifier()) ==
-          state.get_level2().current_names.end())
+        if(!state.get_level2().current_names.has_key(ssa_expr.get_identifier()))
```
Comment (tautschnig) [inline src/goto-symex/auto_objects.cpp]: ~Formatting looks messed up?~ (edit: just a GitHub artifact)

File: src/goto-symex/goto_state.cpp [hunk 2]
```diff
@@ -28,13 +28,22 @@ std::size_t goto_statet::increase_generation(
...
-  return current_emplace_res.first->second.second;
+  if(!r_opt)
+  {
+    level2.current_names.insert(l1_identifier, std::make_pair(lhs, n));
+  }
+  else
+  {
+    std::pair<ssa_exprt, unsigned> copy = r_opt->get();
+    copy.second = n;
+    level2.current_names.replace(l1_identifier, copy);
```
Comment (tautschnig) [inline src/goto-symex/goto_state.cpp]: `std::move(copy)`

File: src/goto-symex/goto_symex_state.cpp [hunk 3]
```diff
@@ -776,16 +776,20 @@ ssa_exprt goto_symex_statet::add_object(
   const irep_idt l0_name = ssa.get_identifier();
   const std::size_t l1_index = index_generator(l0_name);
 
-  // save old L1 name, if any
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp:798]: I'm not sure I appreciate that this comment has been removed.

File: src/goto-symex/goto_symex_state.h [hunk 4]
```diff
@@ -223,15 +223,16 @@ class goto_symex_statet final : public goto_statet
   }
 
   /// Drops an L1 name from the local L2 map
-  void drop_l1_name(symex_renaming_levelt::current_namest::const_iterator it)
+  void drop_existing_l1_name(const irep_idt &l1_identifier)
   {
-    level2.current_names.erase(it);
+    level2.current_names.erase(l1_identifier);
   }
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.h:232]: I think this method should either be removed or else be used from `drop_l1_name`.

File: src/util/ssa_expr.h [hunk 5]
```diff
@@ -16,6 +16,8 @@ Author: Daniel Kroening, kroening@kroening.com
 class ssa_exprt:public symbol_exprt
 {
 public:
+  ssa_exprt() : symbol_exprt("", typet()) {}
```
Comment (tautschnig) [inline src/util/ssa_expr.h]: This is pretty much a show stopper: why is this needed? Presumably a side effect of how objects are created in a sharing map, but can the sharing map please be fixed instead?

**Review by tautschnig — COMMENTED** [review]
Comments (including one blocking one) below; I've started performance experiments, will report on their results later on today.

**Comment by tautschnig** [comment]
The results for my recent favourite ReachSafety-ECA are: 160 points scored with this patch, 156 points without this patch. Symex steps per second increase from 2978.24 to 2991.21. So that's overall an improvement, albeit not a groundbreaking one.

**Comment by romainbrenguier** [comment]
On my benchmarks the total time went from 106.85 to 99.31 chart (2)

**Comment by tautschnig** [comment]
@danpoe @romainbrenguier Could you rerun your benchmarking now that field-sensitivity is merged? It's possible that the maps now actually get larger, and thus such changes may be more impactful.

File: src/goto-symex/goto_symex_state.h [hunk 6]
```diff
@@ -223,15 +223,16 @@ class goto_symex_statet final : public goto_statet
...
 
   /// Drops an L1 name from the local L2 map
   void drop_l1_name(const irep_idt &l1_identifier)
   {
-    level2.current_names.erase(l1_identifier);
+    if(level2.current_names.has_key(l1_identifier))
```
Comment (smowton) [inline src/goto-symex/goto_symex_state.h]: Now have `erase_if_exists`

File: src/goto-symex/renaming_level.cpp [hunk 7]
```diff
@@ -75,18 +78,18 @@ operator()(renamedt<ssa_exprt, L1> l1_expr) const
 
 void symex_level1t::restore_from(const current_namest &other)
 {
-  auto it = current_names.begin();
-  for(const auto &pair : other)
+  current_namest::delta_viewt delta_view;
+  other.get_delta_view(current_names, delta_view, false);
```
Comment (smowton) [inline src/goto-symex/renaming_level.cpp:82]: `delta_view` overestimates the keys that may differ, right? If so couldn't the `replace` below replace for no good reason?

**Review by smowton — APPROVED** [review]
Tested this with a long-running Webgoat analysis -- found before the changes it took ~200s, afterwards ~160s.

**Comment by danpoe** [comment]
Here are some results for the runtime of symex only, with `--unwind 1`, on several methods from Apache Tika, after field-sensitivity was merged:

Rplot

Overall with sharing it's 3% faster.

Btw, I've also compared old symex vs. field sensitive symex and overall field-sensitive symex was 6.3X faster on the benchmarks.

**Comment by romainbrenguier** [comment]
For my benchmarks it didn't seem to have a positive effect:
execution time on develop and sharing renaming (1)
Total time went from 73.86s to 77.218s

**Comment by smowton** [comment]
Now that field-sensitivity is in this makes even more of a difference, mostly in skipping the large number of new keys in the renaming map via a sparse `delta_viewt`. I suspect the benchmarks above aren't really seeing the benefit as they're micro-benchmarks, whereas these effects only really show up once the renaming map gets large.

File: src/util/ssa_expr.h (hunk 5 above)
Comment (tautschnig) [reply src/util/ssa_expr.h]: Maybe somewhere between @danpoe and @smowton: can you look into this and see whether we can get away without it?

Comment (smowton) [reply src/util/ssa_expr.h]: Sure, I'll take a look

Comment (danpoe) [reply src/util/ssa_expr.h]: I was meaning to fix this once we have decided whether to merge this PR. Currently the sharing map creates a dummy object (using the default constructor of the stored value type) when for a `delta_view_itemt` the key is only in one of the maps. We can probably use an `optionalt` instead.

**Comment by danpoe** [comment]
@smowton Would you be able to re-run the benchmarks on the Webgoat analysis now that field-sensitivity has been merged?

File: src/util/ssa_expr.h (hunk 5 above)
Comment (smowton) [reply src/util/ssa_expr.h]: https://github.com/diffblue/cbmc/pull/4612 @danpoe

**Comment by smowton** [comment]
Benchmark results for a long-running Webgoat analysis:

Develop: 77s
#4464: 57s
#4464 + #4505: 26s

File: src/util/ssa_expr.h (hunk 5 above)
Comment (tautschnig) [reply src/util/ssa_expr.h]: #4612 is merged.

**Comment by danpoe** [comment]
Thanks everyone for the benchmarking and comments. So in summary on @romainbrenguier's benchmarks it's about 4% slower, on @smowton's benchmark it's about twice as fast, and on my benchmarks it's about 3% faster. So I'd be inclined to merge it once comments have been addressed and CI passes.

File: src/goto-symex/goto_symex_state.cpp (hunk 3 above)
Comment (danpoe) [reply src/goto-symex/goto_symex_state.cpp:798]: Put back in

File: src/goto-symex/goto_symex_state.h (hunk 4 above)
Comment (danpoe) [reply src/goto-symex/goto_symex_state.h:232]: The other version is now using `erase_if_exists()` as suggested by @smowton

File: src/goto-symex/renaming_level.cpp (hunk 7 above)
Comment (danpoe) [reply src/goto-symex/renaming_level.cpp:82]: Fixed

**Comment by smowton** [comment]
CI failure is caused by the rebase, as `other_m` is now a pointer
